doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,796 @@
+
+/*
+ * Construct and free a new SSL_EVENT object
+ */
+SSL_EVENT *SSL_EVENT_new(void);
+void SSL_EVENT_free(SSL_EVENT *ev);
+
+
+/* Gives an integer value representing the type of event that occurred */
+uint64_t SSL_EVENT_get_event_type(SSL_EVENT *event);
+
+/*
+ * Gets the stream associated with a given event. In the case of "new connection"
+ * this will be NULL. In case the of "new stream" or "connection close" it will
+ * be the "connection" object (i.e. first SSL object for the connection)
+ */
+SSL *SSL_EVENT_get_SSL(SSL_EVENT *event);
+
+/*
+ * Returns the error code associated with the event - for events that have an
+ * associated error code.
+ */
+uint64_t SSL_EVENT_get_error_code(SSL_EVENT *event);
+
+/*
+ * Returns the reason phrase associated with an event if there is one or NULL
+ * otherwise. Only relevant for SSL_EVENT_TYPE_CONNECTION_CLOSE.
+ */
+char *SSL_EVENT_get_reason(SSL_EVENT *event);
+
+/*
+ * Works as normal if the SSL object has already had its BIOs configured.
+ * Otherwise it will check the new connection queue in the SSL_EVENT_CTX. If it
+ * finds a pending new connection then it will "accept" it, adding the BIOs and
+ * running the handshake. In the case of QUIC this will set up the SSL object as
+ * the connection object (with an associated stream if applicable). Will return
+ * SSL_ERROR_WANT_READ in the event that no new conneciton is pending.
+ */
+__owur int SSL_accept(SSL *ssl);
+
+/*
+ * Works as normal for (D)TLS. For QUIC it may be used for connecting new
+ * streams created via `SSL_new_stream()`. For initial connections it may be
+ * used but it is preferrable to use SSL_connect_ex() instead.
+ */
+__owur int SSL_connect(SSL *ssl);
+
+/*
+ * Works as for SSL_connect() except additionally specifying a hostname and
+ * destination address. The (D)TLS and QUIC the hostname sets the hostname that
+ * will be sent in the SNI extension, as well as the hostname used for server
+ * certificate verification. The addr argument specifies for the destination
+ * address to connect to. This will be ignored for (D)TLS. For QUIC it is
+ * mandatory.
+ * For QUIC an SSL object will typically use the BIOs as set on the
+ * SSL_EVENT_CTX rather than having its own BIOs.
+ */
+__owur int SSL_connect_ex(SSL *ssl, const char *hostname, const BIO_ADDR *addr);
+
+/*
+ * In the case of a (D)TLS client attempts a resumption handshake. Will always
+ * fail for a (D)TLS server (returns NULL). For QUIC it creates a new stream
+ * (works on either client or server).
+ * bidi - indicates whether the stream is bidirectional or write only. Selecting
+ *        write only is an error for (D)TLS
+ */
+SSL *SSL_new_stream(SSL *ssl, int bidi);
+
+/*
+ * Not supported for (D)TLS - will always return NULL. For QUIC it will check
+ * if a "new stream" event has occurred for the connection. If it has then a
+ * new SSL object is created to represent it and returned.
+ */
+SSL *SSL_accept_stream(SSL *ssl);
```

Comment:
I've added two new function calls `SSL_CTX_set_params` and `SSL_set_params` for setting arbitrary parameters on an SSL_CTX/SSL object. This could be used (for example) for setting flow control parameters